### PR TITLE
HPCC-27987 Add ability to configure default dfuserver options globally (via environment), e.g. nocommon/maxConnections

### PR DESCRIPTION
**Pull request by AttilaVamos**
Implement
- FileServices.GetNoCommonDefault();
function to read default value from environment.xml in bare-metal system.

Add default values to default environment.xml.in .

Replace noCommon default value with GetNoCommonDefault() function call in:
- SprayFixed
- SprayVariable
- SprayDelimited
- SprayXml
- SprayJson

Add debug log into DFU Server spray to check the value of 'noCommon' and 'maxConnections' parameters.

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:

Tested manually.

### CODE REVIEW

File: plugins/fileservices/fileservices.cpp [hunk 1]
```diff
@@ -3128,3 +3128,49 @@ FILESERVICES_API void FILESERVICES_CALL fsClearExpireDays(ICodeContext * ctx, co
     else
         throw makeStringExceptionV(0, "ClearExpireDays: Could not find logical file %s", lfn.str());
 }
+
+bool getDefaultValue(const char * processName, const char * propertyName, StringBuffer & strDefaultValue)
+{
+    Owned<IConstEnvironment> daliEnv = openDaliEnvironment();
```
Comment (ghalliday) [inline plugins/fileservices/fileservices.cpp:3134]: This shouldn't be called in (is undefined) in containerized version.

File: plugins/fileservices/fileservices.cpp [hunk 2]
```diff
@@ -3128,3 +3128,49 @@ FILESERVICES_API void FILESERVICES_CALL fsClearExpireDays(ICodeContext * ctx, co
...
+    Owned<IPropertyTree> env = getEnvironmentTree(daliEnv);
+
+    if (env.get())
+    {
+        Owned<IPropertyTreeIterator> processIter = env->getElements(processName);
```
Comment (ghalliday) [inline plugins/fileservices/fileservices.cpp:3139]: If these are global defaults, does it make sense to use the values associated with the 1st dfu server?

What is the corresponding implementation for containerized mode?  

(I'm not sure I have a good answer to either question.  @jakesmith ideas?)

File: plugins/fileservices/fileservices.cpp [hunk 3]
```diff
@@ -3128,3 +3128,49 @@ FILESERVICES_API void FILESERVICES_CALL fsClearExpireDays(ICodeContext * ctx, co
...
+    // This function returns the value of the DFUServerProcess' 'noCommon' property, if it is exists. 
+    // Othervise it returns with true, the recent default value of 'noCommon'
+
+    StringBuffer strDefaultValue;
+    if (getDefaultValue("Software/DfuServerProcess", "@noCommon", strDefaultValue))
+        return strToBool(strDefaultValue.str());
+    return true;
+}
+
+FILESERVICES_API int FILESERVICES_CALL fsGetMaxConnectionsDefault()
```
Comment (ghalliday) [inline plugins/fileservices/fileservices.cpp:3163]: Why is this needed, rather than dfuserver using a global default value?  It is already only set in the dfu request if it has been specified.

**Review by ghalliday — CHANGES_REQUESTED** [review]

File: plugins/fileservices/fileservices.cpp (hunk 1 above)
Comment (AttilaVamos) [reply plugins/fileservices/fileservices.cpp:3134]: Yes, that is true. I move it into a conditional block.

File: plugins/fileservices/fileservices.cpp (hunk 2 above)
Comment (AttilaVamos) [reply plugins/fileservices/fileservices.cpp:3139]: My original thought/implementation was that, but then it raised several questions/implications, therefore I moved them from Instance to Server Process.

File: plugins/fileservices/fileservices.cpp (hunk 3 above)
Comment (AttilaVamos) [reply plugins/fileservices/fileservices.cpp:3163]: @jakesmith ?

File: plugins/fileservices/fileservices.cpp [hunk 4]
```diff
@@ -3140,3 +3140,56 @@ FILESERVICES_API void FILESERVICES_CALL fsClearExpireDays(ICodeContext * ctx, co
...
+        // TBD
+
+    #else
+        Owned<IConstEnvironment> daliEnv = openDaliEnvironment();
+        Owned<IPropertyTree> env = getEnvironmentTree(daliEnv);
+
+        if (env.get())
+        {
+            Owned<IPropertyTreeIterator> processIter = env->getElements(processName);
+            if (processIter->first())
```
Comment (jakesmith) [inline plugins/fileservices/fileservices.cpp:3158]: > Gavin Halliday:
> If these are global defaults, does it make sense to use the values associated with the 1st dfu server?
> 

> Attila Vamos:
> My original thought/implementation was that, but then it raised several questions/implications, therefore I moved them from Instance to Server Process.

They should not live in the 1st DfuServerProcess, there may be multiple (not just multiple Instance's under each DfuServerProcess).
The original idea was to just allow these values to be set globally, not per dfuserver.
These settings should move into a global section of the environment (maybe EnvSettings, or could add new Software/Globals section).

File: plugins/fileservices/fileservices.cpp (hunk 3 above)
Comment (jakesmith) [reply plugins/fileservices/fileservices.cpp:3182]: Agree, for this and any other property that can, it should pickup the default within dfuserver when it's processing the dfu workunit.
NB: it already defaults there to a hard-wired DEFAULT_MAX_CONNECTIONS, so that can be switched for the global default if configured.

Unfortunately, the same cannot be applied to boolean flags (e.g. noCommon) where it's default can not be a 'unset' value and therefore cannot be tested in dfuserver in order to use a global default.
For properties that can't (e.g. noCommon), the current approach of picking up a configured global default and passing it in, is the best approach I can think of.

File: initfiles/etc/DIR_NAME/environment.xml.in [hunk 5]
```diff
@@ -217,7 +217,9 @@
                     monitorinterval="900"
                     monitorqueue="dfuserver_monitor_queue"
                     name="mydfuserver"
-                    queue="dfuserver_queue">
+                    queue="dfuserver_queue"
+                    noCommon="true"
+                    maxConnections="-1">
```
Comment (jakesmith) [inline initfiles/etc/DIR_NAME/environment.xml.in:222]: See other comment, should be in a global section of the environment.xml..

File: plugins/fileservices/fileservices.cpp [hunk 6]
```diff
@@ -3140,3 +3140,56 @@ FILESERVICES_API void FILESERVICES_CALL fsClearExpireDays(ICodeContext * ctx, co
...
+    return false;
+}
+
+FILESERVICES_API bool FILESERVICES_CALL fsGetNoCommonDefault()
+{
+    // This function returns the value of the DFUServerProcess' 'noCommon' property, if it is exists. 
+    // Othervise it returns with true, the recent default value of 'noCommon'
+
+    StringBuffer strDefaultValue;
+    if (getDefaultValue("Software/DfuServerProcess", "@noCommon", strDefaultValue))
```
Comment (jakesmith) [inline plugins/fileservices/fileservices.cpp:3177]: this could be called quite often and it's not that cheap (openDaliEnvironment() etc.).
To avoid the cost per call, you could use a singleton pattern to ensure the action is only done on 1st call only.

File: plugins/fileservices/fileservices.cpp [hunk 7]
```diff
@@ -3140,3 +3140,56 @@ FILESERVICES_API void FILESERVICES_CALL fsClearExpireDays(ICodeContext * ctx, co
     else
         throw makeStringExceptionV(0, "ClearExpireDays: Could not find logical file %s", lfn.str());
 }
+
+bool getDefaultValue(const char * processName, const char * propertyName, StringBuffer & strDefaultValue)
+{
+    // Returns with true if a default value retrieved and false if not.
+
+    #ifdef _CONTAINERIZED
```
Comment (jakesmith) [inline plugins/fileservices/fileservices.cpp:3148]: style: we don't normally indent #if's/#endif's

File: plugins/fileservices/fileservices.cpp [hunk 8]
```diff
@@ -3140,3 +3140,56 @@ FILESERVICES_API void FILESERVICES_CALL fsClearExpireDays(ICodeContext * ctx, co
...
+        // TBD
```
Comment (jakesmith) [inline plugins/fileservices/fileservices.cpp:3149]: I think the implementation of this should be trivial.
e.g. in fsGetNoCommonDefault():
```
    return getComponentConfigSP()->getPropBool("@noCommon", true);
```

but should be tested.

**Review by jakesmith — CHANGES_REQUESTED** [review]
@AttilaVamos - please see comments

File: plugins/fileservices/fileservices.cpp (hunk 2 above)
Comment (AttilaVamos) [reply plugins/fileservices/fileservices.cpp]: Changed

File: dali/ft/filecopy.cpp [hunk 9]
```diff
@@ -3087,6 +3087,8 @@ void FileSprayer::spray()
 
     LOG(MCdebugInfo, job, "compressedInput:%d, compressOutput:%d", compressedInput, compressOutput);
     LOG(MCdebugInfo, job, "noCommon:%s", boolToStr(options->getPropBool(ANnocommon)));
+    LOG(MCdebugInfo, job, "maxConnections:%d", options->getPropInt(ANmaxConnections));
+
```
[the remark this answers is not in the document]
Comment (AttilaVamos) [reply dali/ft/filecopy.cpp:3501]: Removed

File: ecllibrary/std/File.ecl [hunk 10]
```diff
@@ -105,6 +105,30 @@ EXPORT FsLandingZoneRecord := lib_fileservices.FsLandingZoneRecord;
...
+    lib_fileservices.FileServices.GetNoCommonDefault();
+
+/**
+ * Returns the integer value of maxConnetcions property of the first DFUServer instance if it is defined,
+ * othervise with -1 as default.
+ *
+ * @return              Integer value of 'maxConnection' property.
+ */
+
+EXPORT INTEGER4 GetMaxCOnnectionsDefault() :=
```
[the remark this answers is not in the document]
Comment (AttilaVamos) [reply ecllibrary/std/File.ecl:125]: Changed.

File: initfiles/etc/DIR_NAME/environment.xml.in (hunk 5 above)
Comment (AttilaVamos) [reply initfiles/etc/DIR_NAME/environment.xml.in:222]: Moved

File: plugins/fileservices/fileservices.cpp (hunk 7 above)
Comment (AttilaVamos) [reply plugins/fileservices/fileservices.cpp:3148]: Changed

File: dali/ft/filecopy.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (jakesmith) [reply dali/ft/filecopy.cpp:3501]: @AttilaVamos - not important in this trivial case of course, but as per code review doc. :https://github.com/hpcc-systems/HPCC-Platform/blob/e33ff93f7bbd4938fafc085f0cadadbec0396b04/devdoc/CodeReviews.md?plain=1#L46
please leave the 'Resolve conversation' button(s) for the reviewers, it helps the reviewer track issues

File: ecllibrary/std/File.ecl [hunk 11]
```diff
@@ -105,6 +105,30 @@ EXPORT FsLandingZoneRecord := lib_fileservices.FsLandingZoneRecord;
...
+ * othervise return with 'true' as default.
+ *
+ * @return              Boolean value of 'noCommon' property.
+ */
+
+EXPORT boolean GetNoCommonDefault() := 
+    lib_fileservices.FileServices.GetNoCommonDefault();
+
+/**
+ * Returns the integer value of maxConnetcions property of the first DFUServer instance if it is defined,
```
[the remark this answers is not in the document]
Comment (AttilaVamos) [reply ecllibrary/std/File.ecl:119]: Changed

File: plugins/fileservices/fileservices.cpp (hunk 8 above)
Comment (AttilaVamos) [reply plugins/fileservices/fileservices.cpp:3149]: Changed

File: plugins/fileservices/fileservices.cpp (hunk 4 above)
Comment (AttilaVamos) [reply plugins/fileservices/fileservices.cpp:3163]: Done

**Review by AttilaVamos — COMMENTED** [review]
Changed

File: ecllibrary/std/File.ecl [hunk 12]
```diff
@@ -105,6 +105,16 @@ EXPORT FsLandingZoneRecord := lib_fileservices.FsLandingZoneRecord;
 
 /*------------------------------------- Spray functions -----------------------------------------------------------*/
 
+/**
+ * Returns the boolean value of 'noCommon' property of the first DFUServer instance if it is defined,
```
Comment (jakesmith) [inline ecllibrary/std/File.ecl:109]: minor: comment is out of date (settings now in Global)

File: initfiles/etc/DIR_NAME/environment.xml.in [hunk 13]
```diff
@@ -20,6 +20,7 @@
         storageWrites="0.0038"/>
  </Hardware>
  <Software>
+  <Globals  noCommon="true" />
```
Comment (jakesmith) [inline initfiles/etc/DIR_NAME/environment.xml.in:23]: trivial: extra space before />

Also, where has maxConnections definition gone?
The comments/discussion said the default should be picked up by dfuserver. Its default value (and any other non-binary) can be better and more easily picked up at the server, and the environment default if configured set then.

File: plugins/fileservices/fileservices.cpp [hunk 14]
```diff
@@ -3140,3 +3140,30 @@ FILESERVICES_API void FILESERVICES_CALL fsClearExpireDays(ICodeContext * ctx, co
...
+
+static int noCommonDef = NotFound;
+FILESERVICES_API bool FILESERVICES_CALL fsGetNoCommonDefault()
+{
+    // This function returns the value of the DFUServerProcess' 'noCommon' property, if it is exists. 
+    // Othervise it returns with true, the recent default value of 'noCommon'
+#ifdef _CONTAINERIZED
+    return getComponentConfigSP()->getPropBool("@noCommon", true);
+#else
+    CriticalSection  noCommonDefCrit;
```
Comment (jakesmith) [inline plugins/fileservices/fileservices.cpp:3152]: This CS can't be a local stack member. It will not protect any other re-entrant call, since they'll all have their own CS they can lock, needs to be a global that any thread coming here locks to lock others out.

File: plugins/fileservices/fileservices.cpp (hunk 4 above)
Comment (jakesmith) [reply plugins/fileservices/fileservices.cpp:3163]: minor: The above code implies there could be multiple Globals sections, which doesn't make a great deal of sense.
Better to change not to iterate..

Comment (jakesmith) [reply plugins/fileservices/fileservices.cpp:3163]: @AttilaVamos - Not sure if you've missed my previous comments, re. please leave the "Resolve conversation" button for use by the reviewers only:
https://github.com/hpcc-systems/HPCC-Platform/pull/16528#discussion_r975491777
https://github.com/hpcc-systems/HPCC-Platform/pull/16528#discussion_r977883966

**Review by jakesmith — CHANGES_REQUESTED** [review]
@AttilaVamos - please see comments.

Also maxConnections functionality seems to have been removed completely?
but the previous comments and discussions were re. picking up the configured default in dfuserver..

File: ecllibrary/std/File.ecl (hunk 12 above)
Comment (AttilaVamos) [reply ecllibrary/std/File.ecl:109]: Changed

File: initfiles/etc/DIR_NAME/environment.xml.in (hunk 13 above)
Comment (AttilaVamos) [reply initfiles/etc/DIR_NAME/environment.xml.in:23]: Removed

File: ecllibrary/std/File.ecl [hunk 15]
```diff
@@ -105,6 +105,16 @@ EXPORT FsLandingZoneRecord := lib_fileservices.FsLandingZoneRecord;
...
+ * othervise return with 'true' as default.
```
Comment (jakesmith) [inline ecllibrary/std/File.ecl:110]: typo: otherwise

File: ecllibrary/std/File.ecl (hunk 12 above)
Comment (ghalliday) [reply ecllibrary/std/File.ecl:109]: @AttilaVamos you marked this as resolved (you shouldn't be doing that as Jake has mentioned), and you haven't changed the comment.

File: initfiles/etc/DIR_NAME/environment.xml.in (hunk 13 above)
Comment (ghalliday) [reply initfiles/etc/DIR_NAME/environment.xml.in:23]: Again you have marked as resolved, but not released the fix.

**Review by ghalliday — CHANGES_REQUESTED** [review]
@AttilaVamos doesn't look like you have addressed the comments from Jake.  Please re-request review once you think it is ready.

File: plugins/fileservices/fileservices.cpp (hunk 14 above)
Comment (AttilaVamos) [reply plugins/fileservices/fileservices.cpp:3152]: Changed

File: ecllibrary/std/File.ecl (hunk 15 above)
Comment (AttilaVamos) [reply ecllibrary/std/File.ecl:110]: Changed

File: dali/ft/filecopy.cpp [hunk 16]
```diff
@@ -2491,6 +2491,37 @@ bool FileSprayer::needToCalcOutput()
     return !usePullOperation() || options->getPropBool(ANverify);
 }
 
+
+CriticalSection  maxConnectionsDefCrit;
```
Comment (jakesmith) [inline dali/ft/filecopy.cpp:2495]: minor: should be a static.

trivial: double space before max.

File: initfiles/etc/DIR_NAME/environment.xml.in [hunk 17]
```diff
@@ -20,6 +20,8 @@
         storageWrites="0.0038"/>
  </Hardware>
  <Software>
+  <Globals  noCommon="true"
```
Comment (jakesmith) [inline initfiles/etc/DIR_NAME/environment.xml.in:23]: trivial: 2 spaces

File: plugins/fileservices/fileservices.cpp [hunk 18]
```diff
@@ -3140,3 +3140,30 @@ FILESERVICES_API void FILESERVICES_CALL fsClearExpireDays(ICodeContext * ctx, co
     else
         throw makeStringExceptionV(0, "ClearExpireDays: Could not find logical file %s", lfn.str());
 }
+
+CriticalSection  noCommonDefCrit;
```
Comment (jakesmith) [inline plugins/fileservices/fileservices.cpp:3144]: minor: should be static

trivial: 2 spaces.

File: dali/ft/filecopy.cpp [hunk 19]
```diff
@@ -2491,6 +2491,37 @@ bool FileSprayer::needToCalcOutput()
...
+static unsigned maxConnectionsDef = NotFound;
+unsigned getMaxConnectionsDefault()
```
Comment (jakesmith) [inline dali/ft/filecopy.cpp:2497]: minor: this could be static

File: dali/ft/filecopy.cpp (hunk 16 above)
Comment (AttilaVamos) [reply dali/ft/filecopy.cpp:2495]: Changed

File: initfiles/etc/DIR_NAME/environment.xml.in (hunk 17 above)
Comment (AttilaVamos) [reply initfiles/etc/DIR_NAME/environment.xml.in:23]: Changed

File: plugins/fileservices/fileservices.cpp (hunk 18 above)
Comment (AttilaVamos) [reply plugins/fileservices/fileservices.cpp:3144]: Changed

File: dali/ft/filecopy.cpp (hunk 19 above)
Comment (AttilaVamos) [reply dali/ft/filecopy.cpp:2497]: Changed

**Review by jakesmith — CHANGES_REQUESTED** [review]
@AttilaVamos - as discussed offline, I think this default (and others if we ever add them), belong in dfuserver (dfurun), not filecopy.
options are passed in from dfuserver through the sprayer it uses, and if set there (if not already set), all places that reference ANmaxConnections will pick up the env. default

File: initfiles/etc/DIR_NAME/environment.xml.in [hunk 20]
```diff
@@ -20,6 +20,8 @@
         storageWrites="0.0038"/>
  </Hardware>
  <Software>
+  <Globals noCommon="true"
+           maxConnections="0"/>
```
Comment (jakesmith) [inline initfiles/etc/DIR_NAME/environment.xml.in:24]: should be -1 as per see file.ecl.

Comment (AttilaVamos) [reply initfiles/etc/DIR_NAME/environment.xml.in]: Changed

File: plugins/fileservices/fileservices.cpp [hunk 21]
```diff
@@ -3140,3 +3140,30 @@ FILESERVICES_API void FILESERVICES_CALL fsClearExpireDays(ICodeContext * ctx, co
     else
         throw makeStringExceptionV(0, "ClearExpireDays: Could not find logical file %s", lfn.str());
 }
+
+static CriticalSection  noCommonDefCrit;
```
Comment (jakesmith) [inline plugins/fileservices/fileservices.cpp:3144]: trivial: still 2 spaces.

File: plugins/fileservices/fileservices.cpp [hunk 22]
```diff
@@ -3140,3 +3140,30 @@ FILESERVICES_API void FILESERVICES_CALL fsClearExpireDays(ICodeContext * ctx, co
...
+    {
+        // Default value if it is not defined in environment.xml
+        noCommonDef = true;
+        Owned<IConstEnvironment> daliEnv = openDaliEnvironment();
+        Owned<IPropertyTree> env = getEnvironmentTree(daliEnv);
+
+        if (env.get())
+        {
+            Owned<IPropertyTree> globals = env->getPropTree("Software/Globals");
+            noCommonDef = globals->getPropBool("@noCommon", noCommonDef);
```
Comment (jakesmith) [inline plugins/fileservices/fileservices.cpp:3164]: trivial: as per code in dfurun, it is unecessary to fetch tree into a linked object, and then the fetch the bool from it. It would be simpler/cleaner to be:
```
    noCommonDef = env->getPropBool("@noCommon", noCommonDef);
```

**Review by jakesmith — APPROVED** [review]
looks ready.

**Comment by jakesmith** [comment]
@ghalliday - please sanity check

File: plugins/fileservices/fileservices.cpp (hunk 21 above)
Comment (AttilaVamos) [reply plugins/fileservices/fileservices.cpp:3144]: Changed

File: plugins/fileservices/fileservices.cpp (hunk 22 above)
Comment (AttilaVamos) [reply plugins/fileservices/fileservices.cpp:3164]: Changed

**Comment by AttilaVamos** [comment]
@ghalliday Should I do anything more for this PR? The users are very keen to use it.

File: dali/dfu/dfurun.cpp [hunk 23]
```diff
@@ -1117,6 +1117,26 @@ class CDFUengine: public CInterface, implements IDFUengine
             opttree->setProp("@encryptKey",encryptkey);
             opttree->setProp("@decryptKey",decryptkey);
         }
+
+        // if maxConnection is not passed as a user option, check for a default defined in the config/environment.
+        if (!opttree->hasProp("@maxConnections"))
+        {
+            int configDefault = -1;
+            if (isContainerized())
```
Comment (ghalliday) [inline dali/dfu/dfurun.cpp:1125]: This is the one situation where you need to use #ifdef _CONTAINERIZED rather than isContainerized().  

The containerized version will not define getEnvironmentFactory(), so you will get link errors/warnings if symbol resolution is enforced at link time.

File: plugins/fileservices/fileservices.cpp [hunk 24]
```diff
@@ -3140,3 +3140,27 @@ FILESERVICES_API void FILESERVICES_CALL fsClearExpireDays(ICodeContext * ctx, co
...
+        // Default value if it is not defined in environment.xml
+        noCommonDef = true;
+        Owned<IConstEnvironment> daliEnv = openDaliEnvironment();
+        Owned<IPropertyTree> env = getEnvironmentTree(daliEnv);
+
+        if (env.get())
+            noCommonDef = env->getPropBool("@noCommon", noCommonDef);
+    }
+#endif
+    return noCommonDef;
```
Comment (ghalliday) [inline plugins/fileservices/fileservices.cpp:3165]: minor: Should explicitly cast to boolean, otherwise you'll get warnings from compilers (especially windows)

File: plugins/proxies/lib_fileservices.ecllib [hunk 25]
```diff
@@ -112,4 +112,5 @@ export FileServices := SERVICE : plugin('fileservices'), time
   integer4 GetExpireDays(const varstring lfn) : c,context,entrypoint='fsGetExpireDays';
   SetExpireDays(const varstring lfn, integer4 expireDays) : c,context,entrypoint='fsSetExpireDays';
   ClearExpireDays(const varstring lfn) : c,context,entrypoint='fsClearExpireDays';
+  boolean GetNoCommonDefault() : c,context,entrypoint='fsGetNoCommonDefault';
```
Comment (ghalliday) [inline plugins/proxies/lib_fileservices.ecllib:115]: I'm not quite sure how this links, probably because it is within an extern "C" so name mangling is disabled.  The function does not have an ICodeContext  * parameter, so it should not have a context attribute in this list.

File: dali/ft/filecopy.cpp [hunk 26]
```diff
@@ -2500,14 +2500,20 @@ unsigned FileSprayer::numParallelConnections(unsigned limit)
...
 unsigned FileSprayer::numParallelSlaves()
 {
-    unsigned numPullers = transferSlaves.ordinality();
+    unsigned numPullers = transferSlaves.ordinality();  // == number of targets
     unsigned maxConnections = DEFAULT_MAX_CONNECTIONS;
     unsigned connectOption = options->getPropInt(ANmaxConnections, 0);
+    LOG(MCdebugInfo, job, "In numParallelSlaves():");
+    LOG(MCdebugInfo, job, "    numPullers:%u", numPullers);
+    LOG(MCdebugInfo, job, "    maxConnections:%u", maxConnections);
+    LOG(MCdebugInfo, job, "    connectOption:%u", connectOption);
```
Comment (ghalliday) [inline dali/ft/filecopy.cpp:2509]: Slightly concerned about the extra logging. does it come out in the dfuserver log by default?  If so it is too verbose.

**Review by ghalliday — CHANGES_REQUESTED** [review]
@AttilaVamos a few problems remaining, and a question.

File: dali/dfu/dfurun.cpp (hunk 23 above)
Comment (AttilaVamos) [reply dali/dfu/dfurun.cpp:1125]: I see. Changed.

File: plugins/fileservices/fileservices.cpp (hunk 24 above)
Comment (AttilaVamos) [reply plugins/fileservices/fileservices.cpp:3165]: Changed

File: plugins/proxies/lib_fileservices.ecllib (hunk 25 above)
Comment (AttilaVamos) [reply plugins/proxies/lib_fileservices.ecllib:115]: Removed

File: dali/ft/filecopy.cpp (hunk 26 above)
Comment (AttilaVamos) [reply dali/ft/filecopy.cpp:2509]: I see. I put them in #ifdef _DEBUG .. #endif block.

**Comment by ghalliday** [comment]
@AttilaVamos please don't squash changes until the review has accepted them.

**Review by ghalliday — APPROVED** [review]
I think it is now ook to merge.
Slightly uneasy about
(1) using Software/Globals rather than copying the global section into the configuration for this component (as is done with costs) and also not cached.
(2) still additional logging.  I think this is can be reviewed as part of a more comprehensive review of logging.

**Comment by ghalliday** [comment]
@AttilaVamos Github has a helpful link to compare the two versions which helped.

**Comment by AttilaVamos** [comment]
@ghalliday Thanks and sorry again.

**Comment by AttilaVamos** [comment]
@richardkchapman please merge.

**Comment by richardkchapman** [comment]
Merged to master and backmerged to 8.10.x for 8.10.0
